### DISCUSSION TITLE
How to builld a plugin

### DISCUSSION THREAD

**Discussion by IvanRodriCalleja** [wooorm/mdxjs-rs#5] [Q&A]
Hi 

I know that at the moment there isn't a plugin system but i am working on a `highlighter` plugin (not public yet). In this moment if i want to run this plugin i have to fork the repo and  add a new phase to the compile function like this: 

<img width="833" alt="image" src="https://user-images.githubusercontent.com/22966198/200837420-35b30f75-cc73-4c0c-8b54-728ac6143d14.png">

In order to build my plugin i need the `hast` definitions to create the needed html code (because i decided to build in this phase) but i need to export in the fork too because now they are not public.

Do you have plans on exporting structs to allow building plugins at html level?

If i publish my crate it won't be possible to use unless you add it in the compile function, is it something you are open to do until a plugin system is implemented?

**Comment by wooorm** [comment]
Cool to hear! :) I don’t know exactly how everything should work and look. So for now I’ve just kept everything private.

Are you using treesitter? Something else?

>  i need the hast definitions

I think exposing `hast` from this project is fine for now. Sure!

> is it something you are open to do until a plugin system is implemented?

We could maybe add an `experimental_hast_transforms` option for now. Should perhaps receive a whole tree (as a value) so it can return something else?

---

I wonder if I should change the ASTs used here to a `Node` trait.
That could make utilities such as https://github.com/wooorm/mdxjs-rs/blob/4a861dcb4a596d8e862a328f053d6c6a53fad76b/src/mdast_util_to_hast.rs#L1197-L1222 work for mdast/hast/etc.

**Reply by IvanRodriCalleja** [reply]
Right now i am just using definitions from https://github.com/wooorm/mdxjs-rs/blob/main/src/hast.rs and i did a `visitor` to find `code` Nodes but it would be nice if this comes from the library.

Let me upload the code i have in a repo just to let you know if it makes sense what i comment before doing anything.

I can help you to implement something if needed.

I'll come to you soon!!

**Reply by IvanRodriCalleja** [reply]
Hi again,

Finally i had some time to spend on the highlighter and upload a initial version with just 2 lexers, you can find [here](https://github.com/IvanRodriCalleja/mdx-rs-highlighter/tree/main/crates/mdx_highlighter). I decided to create as a workspace to have the fork and maybe create my own compile function with the plugin and bindings.

Do you think something like this will work in the future (without needing a fork)?
Right now i only need the fork to pick types and to build my own `compile` function to add my plugin.

**Reply by wooorm** [reply]
Nice! Have you thought about using: https://github.com/tree-sitter/tree-sitter/tree/master/highlight?

My previous comment still stands:
1. Preferably first make nodes into traits so that “unist utilities” can work on mdast and hast
2. We can then expose hast from here or now
3. We can then add some “`experimental_hast_transforms`” or so functionality here

**Reply by IvanRodriCalleja** [reply]
I didn't know about tree-sitter, i'lll take a look.

i'll work on point 1 to make it compatible with both and then i can help you to implement the `experimental_hast_transforms` if you want.

**Comment by n1xx1** [comment]
I just came around this project and the first thing I thought is: "let me build plugins for it now"!

I think having mdast, hast and recma plugins be just functions that return a new tree would be more than enough for a plugin system, similar to unified plugins.

Maybe something like this in the configuration:

```
pub experimental_mdast_transforms: Option<Vec<Box<dyn Fn(&mdast::Node) -> mdast::Node>>>;
pub experimental_hast_transforms: Option<Vec<Box<dyn Fn(&hast::Node) -> hast::Node>>>;
pub experimental_recma_transforms: Option<Vec<Box<dyn Fn(&swc_core::ecma::ast::Module) -> swc_core::ecma::ast::Module>>>;
```

Or maybe even supports both signatures of `Fn(&mut mdast::Node) -> void` and `Fn(&mdast::Node) -> mdast::Node`. But I'm not entirely sure about that. Maybe at some point it could even support wasm plugins like swc? Anyway thanks a lot for this @wooorm!

**Reply by wooorm** [reply]
I'm open to trying something out and experimenting! So whoever had an idea and is interested in working on it, can give it a go!
